### PR TITLE
[INJIMOB-2324]: Adding 3 sec tolerance to dates in Mdoc Validator.

### PR DESCRIPTION
**Pull request by balachandarg-tw**
- Adding 3 sec tolerance to dates in Mdoc Validator.
- Splitted the condition for validFrom/validUntil to return proper error code to help debugging.

Ticket Number:
[INJIMOB-2324](https://mosip.atlassian.net/browse/INJIMOB-2324)

[INJIMOB-2324]: https://mosip.atlassian.net/browse/INJIMOB-2324?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by swatigoel — CHANGES_REQUESTED** [review]

File: vc-verifier/kotlin/vcverifier/src/main/java/io/mosip/vercred/vcverifier/utils/DateUtils.kt [hunk 1]
```diff
@@ -114,9 +114,18 @@ object DateUtils {
...
+            logger.severe("Given date is not available in supported date formats")
+            return false
+        }
+        return inputDate.isInFutureWithTolerance(toleranceInMilliSeconds)
+    }
+
 }
 
-fun Date.isFutureDateWithTolerance(toleranceInMilliSeconds: Long = 3000): Boolean {
+fun Date.isInFutureWithTolerance(toleranceInMilliSeconds: Long = 3000): Boolean {
```
[the remark this answers is not in the document]
Comment (balachandarg-tw) [reply vc-verifier/kotlin/vcverifier/src/main/java/io/mosip/vercred/vcverifier/utils/DateUtils.kt]: Done and updated tests.

File: vc-verifier/kotlin/vcverifier/src/main/java/io/mosip/vercred/vcverifier/credentialverifier/validator/MsoMdocValidator.kt [hunk 2]
```diff
@@ -33,18 +38,29 @@ class MsoMdocValidator {
                 logger.severe("validUntil / validFrom is not available in the credential's MSO")
                 throw ValidationException(ERROR_MESSAGE_INVALID_DATE_MSO, ERROR_CODE_INVALID_DATE_MSO)
             }
-            val isCurrentTimeGreaterThanValidFrom =
-                DateUtils.isDatePassedCurrentDate(validFrom.toString())
-            val isCurrentTimeLessThanValidUntil =
-                !DateUtils.isDatePassedCurrentDate(validUntil.toString())
+            val isValidFromIsFutureDate =
+                DateUtils.isFutureDateWithTolerance(validFrom.toString())
```
Comment (KiruthikaJeyashankar) [inline vc-verifier/kotlin/vcverifier/src/main/java/io/mosip/vercred/vcverifier/credentialverifier/validator/MsoMdocValidator.kt:42]: validFrom attribute is expected to be before current DateTime, the function looks like we are checking if validFrom is future date (validFrom < currentTime). Is this expected here?

Comment (balachandarg-tw) [reply vc-verifier/kotlin/vcverifier/src/main/java/io/mosip/vercred/vcverifier/credentialverifier/validator/MsoMdocValidator.kt:42]: Yes, validFrom should come before current date, so when validFrom is Future date , we are throwing exception.
```
            if(isValidFromIsFutureDate){
                logger.severe("Error while doing validity verification - invalid validFrom in the MSO of the credential")
                throw ValidationException(ERROR_MESSAGE_INVALID_VALID_FROM_MSO, ERROR_CODE_INVALID_VALID_FROM_MSO)
            }
```

File: vc-verifier/kotlin/vcverifier/src/test/java/io/mosip/vercred/vcverifier/credentialverifier/validator/MsoMdocValidatorTest.kt [hunk 3]
```diff
@@ -49,27 +51,27 @@ class MsoMdocValidatorTest {
...
         }
 
-        assertEquals(ERROR_MESSAGE_INVALID_DATE_MSO,verificationException.errorMessage)
-        assertEquals(ERROR_CODE_INVALID_DATE_MSO,verificationException.errorCode)
+        assertEquals(ERROR_MESSAGE_INVALID_VALID_UNTIL_MSO,verificationException.errorMessage)
+        assertEquals(ERROR_CODE_INVALID_VALID_UNTIL_MSO,verificationException.errorCode)
     }
 
     @Test
```
Comment (KiruthikaJeyashankar) [inline vc-verifier/kotlin/vcverifier/src/test/java/io/mosip/vercred/vcverifier/credentialverifier/validator/MsoMdocValidatorTest.kt:77]: We have now split the error for date checks of mso, can we also add respective unit tests for the same?

**Review by swatigoel — APPROVED** [review]

**Pull request merged**